### PR TITLE
Support for microprofile context propagation

### PR DESCRIPTION
**Pull request by tobiasstadler**

### CODE REVIEW

**Comment by Ladicek** [comment]
I don't know much about OpenTracing and Context Propagation, but would like to note that SmallRye OpenTracing is used in projects (Thorntail, WildFly) that don't use Context Propagation.

**Comment by pavolloffay** [comment]
I would like to also understand the motivation why this should be a hard requirement? Perhaps it can be added as a separate module.

**Comment by kenfinnigan** [comment]
+1

@tobiasstadler can you provide some details on the use case this is needed for?

**Comment by tobiasstadler** [comment]
My application receives requests which executes a couple of tasks (rule evaluation) in parallel via MannagedExecutorService. The request thread waits until all tasks are finished (no rule fired) or one of the tasks returns a special value (rule fired). The request and the tasks are traced. The spans of the tasks should be children of the span of the request.

Since spans are thread local I need to propagate them to the threads of the tasks.

**Comment by kenfinnigan** [comment]
Thoughts @pavolloffay?

**Comment by pavolloffay** [comment]
As the MP context propagation is not in the umbrella spec and it's also not required by MP-OpenTracing we shouldn't pull it in by default.

I would rather introduce it as a separate module or as an optional dependency.

**Comment by tobiasstadler** [comment]
Ok, will you create a new module under the small rye organization or should I host it my self?

**Comment by kenfinnigan** [comment]
I think @pavolloffay just meant a new module within this repository.

Something like `context-propagation` at the same level as `implementation` in this repository.

That's my interpretation anyway

**Comment by tobiasstadler** [comment]
@kenfinnigan I created a new module for the context propagation support

File: context-propagation/pom.xml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <parent>
+    <artifactId>smallrye-opentracing-parent</artifactId>
+    <groupId>io.smallrye</groupId>
+    <version>1.3.3-SNAPSHOT</version>
```
Comment (pavolloffay) [inline context-propagation/pom.xml]: Could you please update this? We have released a new version

File: context-propagation/src/main/java/io/smallrye/opentracing/SmallRyeOpenTracingContextProvider.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                };
+            }
+
+            return () -> {
+            };
+        };
+    }
+
+    @Override
+    public ThreadContextSnapshot clearedContext(Map<String, String> props) {
```
Comment (pavolloffay) [inline context-propagation/src/main/java/io/smallrye/opentracing/SmallRyeOpenTracingContextProvider.java]: I am not familiar with this API. Could you please comment why we don't have to handle OT context here?

File: context-propagation/src/main/java/io/smallrye/opentracing/SmallRyeOpenTracingContextProvider.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import io.opentracing.Scope;
+import io.opentracing.Span;
+import io.opentracing.Tracer;
+import io.opentracing.util.GlobalTracer;
+
+public class SmallRyeOpenTracingContextProvider implements ThreadContextProvider {
+
+    @Override
+    public ThreadContextSnapshot currentContext(Map<String, String> props) {
+        Tracer activeTracer = GlobalTracer.get();
```
Comment (pavolloffay) [inline context-propagation/src/main/java/io/smallrye/opentracing/SmallRyeOpenTracingContextProvider.java]: nit: move this to a final static class field.

File: context-propagation/pom.xml (hunk 1 above)
Comment (tobiasstadler) [reply context-propagation/pom.xml]: done

File: context-propagation/src/main/java/io/smallrye/opentracing/SmallRyeOpenTracingContextProvider.java (hunk 3 above)
Comment (tobiasstadler) [reply context-propagation/src/main/java/io/smallrye/opentracing/SmallRyeOpenTracingContextProvider.java]: currentContext is called, when submitting a task to a ManagedExecutorService, so currentContext my be called in parallel by multiple threads. I don't think putting it an (volatile) static field is the best solution

File: context-propagation/src/main/java/io/smallrye/opentracing/SmallRyeOpenTracingContextProvider.java (hunk 2 above)
Comment (tobiasstadler) [reply context-propagation/src/main/java/io/smallrye/opentracing/SmallRyeOpenTracingContextProvider.java]: I tried implementing removing the active span from the propagated context

**Comment by tobiasstadler** [comment]
I just noticed that smallrye/smallrye-context-propagation has modules for different context propagation providers. Should I submit my implementation there?

**Comment by pavolloffay** [comment]
Question for @kenfinnigan ? ^^

**Comment by kenfinnigan** [comment]
It would be interesting to get the thoughts of @FroMage and @manovotn, but from looking at the modules in SRye Context Propagation and that they implement `ThreadContextProvider` as this PR does, I'm leaning towards it makes more sense to go in that repo instead.

**Comment by manovotn** [comment]
SmallRye CP repo contains impl mainly for contexts that are defined by the spec (cdi, JTA). But there is rx-java too.
I think it really depends on where you want your users to see that there is an option for CP integration. Personally, I think it'd be nice to have them under SR CP repo so long as it is an optional module.

Also, I'm wondering what's the situation with the new FT impl - I recall @michalszynkiewicz asking about CP usage there. Do you have a module for it or is it baked right into the impl?

File: context-propagation/src/main/java/io/smallrye/opentracing/SmallRyeOpenTracingContextProvider.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+        return () -> () -> {
+            activeTracer.scopeManager().activate(null, false);
+        };
+    }
+
+    @Override
+    public String getThreadContextType() {
+        return "OPENTRACING_CONTEXT";
```
Comment (manovotn) [inline context-propagation/src/main/java/io/smallrye/opentracing/SmallRyeOpenTracingContextProvider.java]: Maybe just `OPENTRACING`?
Look at spec, none of the contexts there have the `context` suffix.

File: context-propagation/src/main/java/io/smallrye/opentracing/SmallRyeOpenTracingContextProvider.java (hunk 2 above)
Comment (manovotn) [reply context-propagation/src/main/java/io/smallrye/opentracing/SmallRyeOpenTracingContextProvider.java]: `clearedContext` is for when user specifically wants given context *not* to propagate.
Now, I might be wrong as I don't know much about opentracing, but it looks like you are propagating it nonetheless?

**Comment by Ladicek** [comment]
The ConProp integration in FT is a separate module (and a ServiceLoader mechanism), it's not baked into the core.

**Comment by manovotn** [comment]
Thanks for the info @Ladicek. That means we are already split on where to put these integrations :-/

**Comment by Ladicek** [comment]
I think it's inevitable for ConProp integrations to live outside of SRye ConProp. For example, I've had @ochaloup mentioning he'd like to add one to Narayana :-)

**Comment by manovotn** [comment]
I suppose you are right; especially given that some projects, such as Quarkus, will need a more "customized" integrations. We already have one CDI integration in SR CP (for Weld) and one another in Quarkus (for ArC).

**Comment by manovotn** [comment]
Seeing this PR is hanging around for some time now, is there anything blocking it @pavolloffay ?

**Comment by Ladicek** [comment]
For your consideration, this is how we implement ConProp for OpenTracing in SRye Fault Tolerance: https://github.com/smallrye/smallrye-fault-tolerance/blob/master/implementation/tracing-propagation/src/main/java/io/smallrye/faulttolerance/tracing/TracingContextProvider.java

There's one TODO in the code that should be resolved, but even with that, it's more complete than what's in this PR.

**Comment by manovotn** [comment]
@Ladicek so you are saying this is a duplicate of an already existing and functional implementation?

**Comment by Ladicek** [comment]
More or less. I have already pointed out the existence of OpenTracing ConProp in SRye FT here: https://github.com/smallrye/smallrye-opentracing/pull/103#issuecomment-581298785

I actually think it would be best if the ConProp integration for OpenTracing existed in SRye OpenTracing, but I don't really mind :-)

**Comment by manovotn** [comment]
Oops my bad, back there I thought you are talking about CP of something else.

I agree that SR OpenTracing is a better place for it to reside in.
So, do you think we should just take the FT code and port it here? Since it's already working and being used anyway, it sounds better than writing the same code ground up, WDYT?

**Comment by Ladicek** [comment]
If the one TODO in that code could be resolved, that would be great, but other than that, yea, that's what I would suggest :-)

**Comment by radcortez** [comment]
OpenTracing is discontinued, so SmallRye OpenTracing is in maintenance mode until we do a complete transition to OpenTelemetry. For this reason, we are not adding new features.

**Pull request closed without merging**